pkg/apis/core/validation/validation.go

```diff
@@ -7253,3 +7274,75 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(podAffinityTerm.NamespaceSelector, labelSelectorValidationOptions, fldPath.Child("namespaceSelector"))...)
 	return allErrs
 }
+
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node selector was nil, anything can be set.
+	if oldNodeSelector == nil {
+		return allErrs
+	}
+	// Validate no existing node selectors were deleted or mutated.
+	for k, v1 := range oldNodeSelector {
+		if v2, ok := newNodeSelector[k]; !ok || v1 != v2 {
+			allErrs = append(allErrs, field.Invalid(fldPath, newNodeSelector, "only additions to spec.nodeSelector are allowed (no mutations or deletions)"))
+			return allErrs
+		}
+	}
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
```

Comment:
we should validate that the number of old and new NodeSelectorTerms are the same, no additions to a non-zero length NodeSelectorTerms is allowed because semantically it is an OR.